### ISSUE TITLE
Support multi-file smart contracts (debugging)

### ISSUE THREAD

**Issue by csmuller** [neow3j/neow3j#185]
The compiler should support smart contracts that are spread over multiple `.java` files.

Because Java does not have functionality like C# provides with the keyword `partial`, another solution is needed for enabling multi-file contracts.  
Our proposal is to use an annotation which has a name or ID attribute. That annotation is then applied to every class that is part of the smart contract and the compiler can determine using the name/ID of that annotation if two or more files belong to the same contract.

**Comment by gsmachado** [comment]
As discussed in #242, we need to discuss (and do) the following:

There are two things to be discussed, still:

- [ ] When the method `.compileClasses(canonicalClassNames, setOfSourceSetFilesPath)` is used, meaning that we give several Smart Contract classes to be compiled, as well as with multiple sources, shouldn't we return also multiple `CompilationUnit` objects?
- [ ] If the answer from the previous question is "no, just one `CompilationUnit` is required", then, how should we infer the NEF, Debug Info, and Manifest file names?

We should have multiple ways to specify Smart Contract class names to be compiled:

For multiple files:

```
neow3jCompiler {
    classNames "io.neow3j.examples.NameService", "io.neow3j.examples.NEP5"
    debug false
}
```

For single file:

```
neow3jCompiler {
    classNames "io.neow3j.examples.NameService"
    debug false
}
```

For multiple files with array:

```
neow3jCompiler {
    classNames = [ "io.neow3j.examples.NameService", "io.neow3j.examples.NEP5" ]
    debug false
}
```

For single file with array:

```
neow3jCompiler {
    classNames = [ "io.neow3j.examples.NameService" ]
    debug false
}
```

For single file, keeping backwards compatibility:

```
neow3jCompiler {
    className = "io.neow3j.examples.NameService"
    debug = false
}
```

**Comment by csmuller** [comment]
I challenge the option I mentioned in this issue's description. I don't anymore consider it a good option to have a contract spread it's public methods over multiple class files. That does not seem intuitive at all. There should only be one main contract class file that holds all public (invokable) contract methods.  

That doesn't mean that all contract functionality has to be in the same file. The compiler should still support multiple files (and classes). The files besides the main contract class contain classes that encapsulate parts of the contract's logic. Custom objects used in the contract for organizing data can also live in their own files.

The step to smart contract libraries support isn't far from here.

**Comment by gsmachado** [comment]
> That does not seem intuitive at all. There should only be one main contract class file that holds all public (invokable) contract methods.
> 

Ok, I agree with this argument: so, the definition of a smart contract in Java, is:
"a .java file which contains, among other information, all 'invokable' methods which users might interact with"

> That doesn't mean that all contract functionality has to be in the same file. The compiler should still support multiple files (and classes). The files besides the main contract class contain classes that encapsulate parts of the contract's logic. Custom objects used in the contract for organizing data can also live in their own files.
> 

Yes... so, it means that: all other .java files are "dependencies" of the smart contract, and should be also compiled and "embedded" to the final `.avm`. Right?

> The step to smart contract libraries support isn't far from here.

Absolutely. I also think that we're not far from it! 🚀  ⭐

**Comment by gsmachado** [comment]
@csmuller 

Therefore, based on [this comment](https://github.com/neow3j/neow3j/issues/185#issuecomment-719651582), I think it would make sense to do the following:

```
neow3jCompiler {
    classNames = [ "io.neow3j.examples.NameService" ]
    debug false
}
```

The `className` or `classNames` attribute should contain the "invokable" smart contract class name... and not all the dependencies of that smart contract. That's how I think about it, ideally.

However, nothing prevents us to start small: maybe we can list all the classes, and automatically identify which class is the "invokable" one.

What do you think?

**Comment by csmuller** [comment]
> The `className` or `classNames` attribute should contain the "invokable" smart contract class name... and not all the dependencies of that smart contract. That's how I think about it, ideally.

Yes, i was imagining the same. The contract class is like the entry point to the compilation and the other classes will be processed because the contract class makes use of them.

> However, nothing prevents us to start small: maybe we can list all the classes, and automatically identify which class is the "invokable" one.

I'm not a huge fan of automatically identifying it because I think the developer gets a clearer picture if she actually has to state which class is the contract class.

**Comment by csmuller** [comment]
What I'm not completely sure about yet is the discovery of the source files.
I think the Gradle plugin can search through the whole Gradle project (the smart contract project), collect all .java files and pass them on to the `Compiler`. The compiler can then resolve the mapping between .java files and .class files by reading the `sourceFile` field variable on the ASM `ClassNode`.

The only problem with that approach is, that the `sourceFile` field only holds the last part of the file and not the path that would reveal the fully qualified name. Thus, there could be name collisions.

**Comment by csmuller** [comment]
Solved in #507.
